ksqldb-engine/src/main/java/io/confluent/ksql/physical/scalablepush/consumer/CatchupConsumer.java

```diff
@@ -0,0 +1,248 @@
+   * If it's being called manually, a mapping of starting offsets is given.  This ensures that we
+   * take our current position to be the one we seek to rather than the last committed offsets.
+   * @param startingOffsets The starting offsets to use, if any.  Otherwise, we use the last
+   *     committed.
+   */
+  protected void onNewAssignment(final Optional<Map<Integer, Long>> startingOffsets) {
+    final Set<TopicPartition> tps = waitForNewAssignmentFromLatestConsumer();
+
+    consumer.assign(tps);
+    updateCurrentPositions(startingOffsets);
+    newAssignment = false;
+  }
+
+  /**
+   * Called when a new assignment has been made with a call to {@link #newAssignment(Collection)}.
+   * Assumes it should take the current position to be the last committed offsets.
+   */
+  @Override
+  protected void onNewAssignment() {
+    onNewAssignment(Optional.empty());
+  }
+
+  /**
+   * The first time we start up a LatestConsumer, if it doesn't already have an assignment, we need
+   * to wait for it to give us that assignment by calling {@link #newAssignment}
+   * @return The assignment given by latest
+   */
+  protected Set<TopicPartition> waitForNewAssignmentFromLatestConsumer() {
+    final long startMs = clock.millis();
+    Set<TopicPartition> tps = this.topicPartitions.get();
+    while (tps == null) {
+      if (clock.millis() - startMs > WAIT_FOR_ASSIGNMENT_MS) {
+        throw new KsqlException("Timed out waiting for assignment from Latest");
+      }
+      sleepMs.accept(100L);
+      tps = this.topicPartitions.get();
+    }
+    return tps;
+  }
+
+  @Override
+  protected void subscribeOrAssign() {
+    // We should always have a latest consumer running
+    final LatestConsumer latestConsumer = latestConsumerSupplier.get();
+    Preconditions.checkNotNull(latestConsumer,
+        "Latest should always be started before catchup is run");
+    // Assign current offsets from the latest consumer.  If the latest consumer doesn't have its
+    // assignment yet, this might be null, but we then count on the latest updating this consumer
+    // with the assignment when it's received.
+    this.newAssignment(latestConsumer.getAssignment());
+    final Map<Integer, Long> startingOffsets
+        = offsetRange.getEndOffsets().getSparseRepresentation();
+    onNewAssignment(Optional.of(startingOffsets));
+    // Seek to the provided starting offsets
+    for (TopicPartition tp : consumer.assignment()) {
+      if (startingOffsets.containsKey(tp.partition()) && startingOffsets.get(tp.partition()) >= 0) {
+        consumer.seek(tp, startingOffsets.get(tp.partition()));
+      }
+    }
+  }
+
+  protected void afterOfferedRow(final ProcessingQueue processingQueue) {
+    // Since we handle only one request at a time, we can afford to block and wait for the request.
+    while (processingQueue.isAtLimit()) {
+      LOG.info("Sleeping for a bit since queue is full");
+      sleepMs.accept(1000L);
```

Comment:
What do you think about sleeping for a much shorter amount of time? It seems like this amount of pausing could put us into a sawtooth pattern, where we process a couple of batches, fill up the queue, then the consumer drains the queue a couple of ms later but we sit idle for a full second, then full up the queue again, etc.